de.metas.payment.esr/src/main/java/de/metas/payment/esr/dataimporter/impl/camt54/ESRDataImporterCamt54.java

```diff
@@ -124,294 +100,111 @@ public ESRDataImporterCamt54(@NonNull final I_ESR_Import header, @NonNull final
 			final XMLStreamReader xsr = xif.createXMLStreamReader(input);
-
-			final JAXBContext context = JAXBContext.newInstance(Document.class);
-			final Unmarshaller unmarshaller = context.createUnmarshaller();
-
+	
 			// https://github.com/metasfresh/metasfresh/issues/1903
 			// use a delegate to make sure that the unmarshaller won't refuse camt.054.001.04 and amt.054.001.05
-			@SuppressWarnings("unchecked")
-			final JAXBElement<Document> e = (JAXBElement<Document>)unmarshaller.unmarshal(new MyStreamReaderDelegate(xsr));
-
-			document = e.getValue();
-		}
-		catch (final JAXBException | XMLStreamException e)
-		{
-			throw AdempiereException.wrapIfNeeded(e);
-		}
-
-		final BankToCustomerDebitCreditNotificationV06 bkToCstmrDbtCdtNtfctn = document.getBkToCstmrDbtCdtNtfctn();
-		return bkToCstmrDbtCdtNtfctn;
-	}
-
-	/**
-	 * Thanks to <a href="https://stackoverflow.com/a/6747695/1012103">https://stackoverflow.com/a/6747695/1012103</a>
-	 *
-	 * @author metas-dev <dev@metasfresh.com>
-	 * @task https://github.com/metasfresh/metasfresh/issues/1903
-	 *
-	 */
-	private static class MyStreamReaderDelegate extends StreamReaderDelegate
-	{
-		public MyStreamReaderDelegate(@NonNull final XMLStreamReader xsr)
-		{
-			super(xsr);
-		}
-
-		/**
-		 * If the actual namespace is <code>urn:iso:std:iso:20022:tech:xsd:camt.054.001.04</code> or <code>...001.05</code>, then this method returns <code>...001.06</code> instead,
-		 * because it works for the XML we process here, and it allows us to also process older XML files.
-		 *
-		 * @task https://github.com/metasfresh/metasfresh/issues/1903
-		 */
-		@Override
-		public String getNamespaceURI()
-		{
-			final String namespaceURI = super.getNamespaceURI();
-			if (Objects.equal("urn:iso:std:iso:20022:tech:xsd:camt.054.001.04", namespaceURI)
-					|| Objects.equal("urn:iso:std:iso:20022:tech:xsd:camt.054.001.05", namespaceURI))
+			final MyStreamReaderDelegate mxsr = new MyStreamReaderDelegate(xsr);
+			
+			if (isVersion2Schema(getNameSpaceURI(mxsr)))
 			{
-				// listing those two URNs that we replace is not elegant, but simple & easy. that's why I do it.
-				return "urn:iso:std:iso:20022:tech:xsd:camt.054.001.06";
-			}
-
-			return namespaceURI;
-		}
+				final ESRDataImporterCamt54v02 importerV02 = new ESRDataImporterCamt54v02(header, mxsr);
+				final BankToCustomerDebitCreditNotificationV02 bkToCstmrDbtCdtNtfctn = importerV02.loadXML();
+				if (bkToCstmrDbtCdtNtfctn.getGrpHdr() != null && bkToCstmrDbtCdtNtfctn.getGrpHdr().getAddtlInf() != null)
+				{
+					Loggables.get().withLogger(logger, Level.INFO).addLog("The given input is a test file: bkToCstmrDbtCdtNtfctn/grpHdr/addtlInf={}", bkToCstmrDbtCdtNtfctn.getGrpHdr().getAddtlInf());
+				}
+				
+				try (final IAutoCloseable switchContext = Env.switchContext(InterfaceWrapperHelper.getCtx(header, true)))
+				{
+					return importerV02.createESRStatement(bkToCstmrDbtCdtNtfctn);
+				}
 
-	}
+			}
+			else
```

Comment:
Pelase don't assume that it must be 06 or 02. It might be none of both.

Reply:
We support 02, 04,05 and 06. But 04 and 05 are dealt with 06. So.. from there.. if is not 02, go with 06 (05, 04) version. If is not one of those, will fail.

Reply:
ok